This project is part of the Business Analytics course. The data used for PCA is image data on the types of diseases in tomato plants from Kaggle https://www.kaggle.com/kaustubhb999/tomatoleaf.
This data only contains three types of tomato plant diseases, namely diseases caused by Mosaic Virus, Yellow Leaf Curl Virus, and disease Septoria Leaf Spot, with 10 photos each for every disease in th preliminary data. These three types of disease are represented by spots or shapes which is unusual on the leaves.

The aim of this project is to experiment with many techniques to boost up the performance of the classifier. At first, we used Naive Bayes for the classifier as a performance standard. The techniques are as follows:

1. Add more train data
2. Add more PC features
3. Do feature scaling
4. Change model algorithm
5. Do hyperparameter tuning




# Hypothesis
My initial hypothesis was that the low accuracy was due to too little data for training which was only 10 pieces for each type of disease. While the types of diseases that have a fairly high similarity, for example, leaves infected with Tomato Virus Mosaics and Tomato Yellow Leaf Curl Virus which both have signs of leaves turning bright yellow, the intensity of yellow in the Tomato Yellow Leaf Curl Virus is higher. So, I think the model needs more data to learn.

# Naive Bayes Model for the preliminary data which has 30 rows


Without Feature Scaling

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30baris.csv")

In [ ]:
x = df.drop(['class'], axis=1)
y = df['class']

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 123)

In [ ]:
from sklearn.naive_bayes import GaussianNB
# Mengaktifkan/memanggil/membuat fungsi klasifikasi Naive Bayes 
modelnb = GaussianNB()
# Memasukkan data training pada fungsi klasifikasi Naive Bayes
nbtrain = modelnb.fit(x_train, y_train)

In [ ]:
#prediksi dengan x_test pada data train
y_predict = modelnb.predict(x_test)

print(y_predict)

['TomatoYellowLeafCurlVirus' 'TomatoMosaicVirus'
 'TomatoYellowLeafCurlVirus' 'TomatoMosaicVirus' 'TomatoMosaicVirus'
 'TomatoMosaicVirus']


In [ ]:
#Untuk data yang telah distandarisasi
y_predict = modelnb.predict(x_test)

print(y_predict)

In [ ]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_predict)

array([[0, 0, 0],
       [3, 0, 0],
       [1, 0, 2]])

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predict))

                           precision    recall  f1-score   support

        TomatoMosaicVirus       0.00      0.00      0.00         0
   TomatoSeptoriaLeafSpot       0.00      0.00      0.00         3
TomatoYellowLeafCurlVirus       1.00      0.67      0.80         3

                 accuracy                           0.33         6
                macro avg       0.33      0.22      0.27         6
             weighted avg       0.50      0.33      0.40         6



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


With Feature Scaling 

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30baris.csv")

x = df.drop(['class'], axis=1)
y = df['class']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 123)

#melakukan standarisasi
from sklearn import preprocessing
std_scale = preprocessing.StandardScaler().fit(x_train)
x_train_std = std_scale.transform(x_train)
x_test_std = std_scale.transform(x_test)

#Untuk data yang telah distandarisasi
from sklearn.naive_bayes import GaussianNB
# Mengaktifkan/memanggil/membuat fungsi klasifikasi Naive Bayes 
modelnb = GaussianNB()
# Memasukkan data training pada fungsi klasifikasi Naive Bayes
nbtrain = modelnb.fit(x_train_std, y_train)

#prediksi dengan x_test_std pada data train
y_predict = modelnb.predict(x_test_std)

print(y_predict)

['TomatoYellowLeafCurlVirus' 'TomatoMosaicVirus'
 'TomatoYellowLeafCurlVirus' 'TomatoMosaicVirus' 'TomatoMosaicVirus'
 'TomatoMosaicVirus']


In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predict))

                           precision    recall  f1-score   support

        TomatoMosaicVirus       0.00      0.00      0.00         0
   TomatoSeptoriaLeafSpot       0.00      0.00      0.00         3
TomatoYellowLeafCurlVirus       1.00      0.67      0.80         3

                 accuracy                           0.33         6
                macro avg       0.33      0.22      0.27         6
             weighted avg       0.50      0.33      0.40         6



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


#Naive Bayes Model with more data which has 60 rows

Without Feature Scaling

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 123)

from sklearn.naive_bayes import GaussianNB
# Mengaktifkan/memanggil/membuat fungsi klasifikasi Naive Bayes 
modelnb = GaussianNB()
# Memasukkan data training pada fungsi klasifikasi Naive Bayes
nbtrain = modelnb.fit(x_train, y_train)

#prediksi dengan x_test pada data train
y_predict = modelnb.predict(x_test)

print(y_predict)

['TomatoSeptoriaLeafSpot' 'TomatoYellowLeafCurlVirus'
 'TomatoSeptoriaLeafSpot' 'TomatoYellowLeafCurlVirus'
 'TomatoSeptoriaLeafSpot' 'TomatoMosaicVirus' 'TomatoSeptoriaLeafSpot'
 'TomatoMosaicVirus' 'TomatoMosaicVirus' 'TomatoSeptoriaLeafSpot'
 'TomatoYellowLeafCurlVirus' 'TomatoYellowLeafCurlVirus']


In [ ]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_predict)

array([[3, 1, 0],
       [0, 3, 0],
       [0, 1, 4]])

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predict))

                           precision    recall  f1-score   support

        TomatoMosaicVirus       1.00      0.75      0.86         4
   TomatoSeptoriaLeafSpot       0.60      1.00      0.75         3
TomatoYellowLeafCurlVirus       1.00      0.80      0.89         5

                 accuracy                           0.83        12
                macro avg       0.87      0.85      0.83        12
             weighted avg       0.90      0.83      0.84        12



With Feature Scaling

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 123)

#melakukan standarisasi
from sklearn import preprocessing
std_scale = preprocessing.StandardScaler().fit(x_train)
x_train_std = std_scale.transform(x_train)
x_test_std = std_scale.transform(x_test)

#Untuk data yang telah distandarisasi
from sklearn.naive_bayes import GaussianNB
# Mengaktifkan/memanggil/membuat fungsi klasifikasi Naive Bayes 
modelnb = GaussianNB()
# Memasukkan data training pada fungsi klasifikasi Naive Bayes
nbtrain = modelnb.fit(x_train_std, y_train)

#prediksi dengan x_test_std pada data train
y_predict = modelnb.predict(x_test_std)

print(y_predict)

['TomatoSeptoriaLeafSpot' 'TomatoYellowLeafCurlVirus'
 'TomatoSeptoriaLeafSpot' 'TomatoYellowLeafCurlVirus'
 'TomatoSeptoriaLeafSpot' 'TomatoMosaicVirus' 'TomatoSeptoriaLeafSpot'
 'TomatoMosaicVirus' 'TomatoMosaicVirus' 'TomatoSeptoriaLeafSpot'
 'TomatoYellowLeafCurlVirus' 'TomatoYellowLeafCurlVirus']


In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predict))

                           precision    recall  f1-score   support

        TomatoMosaicVirus       1.00      0.75      0.86         4
   TomatoSeptoriaLeafSpot       0.60      1.00      0.75         3
TomatoYellowLeafCurlVirus       1.00      0.80      0.89         5

                 accuracy                           0.83        12
                macro avg       0.87      0.85      0.83        12
             weighted avg       0.90      0.83      0.84        12



# Naive Bayes Model with adding more PC (75 PC)

For 60 rows data

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60barisPC75.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 123)

from sklearn.naive_bayes import GaussianNB
# Mengaktifkan/memanggil/membuat fungsi klasifikasi Naive Bayes 
modelnb = GaussianNB()
# Memasukkan data training pada fungsi klasifikasi Naive Bayes
nbtrain = modelnb.fit(x_train, y_train)

#prediksi dengan x_test pada data train
y_predict = modelnb.predict(x_test)

print(y_predict)



['TomatoYellowLeafCurlVirus' 'TomatoSeptoriaLeafSpot' 'TomatoMosaicVirus'
 'TomatoSeptoriaLeafSpot' 'TomatoMosaicVirus' 'TomatoSeptoriaLeafSpot'
 'TomatoMosaicVirus' 'TomatoMosaicVirus' 'TomatoSeptoriaLeafSpot'
 'TomatoSeptoriaLeafSpot' 'TomatoYellowLeafCurlVirus'
 'TomatoYellowLeafCurlVirus']


In [ ]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_predict)

array([[2, 1, 0],
       [2, 2, 0],
       [0, 2, 3]])

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predict))

                           precision    recall  f1-score   support

        TomatoMosaicVirus       0.50      0.67      0.57         3
   TomatoSeptoriaLeafSpot       0.40      0.50      0.44         4
TomatoYellowLeafCurlVirus       1.00      0.60      0.75         5

                 accuracy                           0.58        12
                macro avg       0.63      0.59      0.59        12
             weighted avg       0.67      0.58      0.60        12



For 30 rows data

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30barisPC75.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 123)

from sklearn.naive_bayes import GaussianNB
# Mengaktifkan/memanggil/membuat fungsi klasifikasi Naive Bayes 
modelnb = GaussianNB()
# Memasukkan data training pada fungsi klasifikasi Naive Bayes
nbtrain = modelnb.fit(x_train, y_train)

#prediksi dengan x_test pada data train
y_predict = modelnb.predict(x_test)

print(y_predict)


['TomatoSeptoriaLeafSpot' 'TomatoMosaicVirus' 'TomatoYellowLeafCurlVirus'
 'TomatoMosaicVirus' 'TomatoSeptoriaLeafSpot' 'TomatoMosaicVirus']


In [ ]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, y_predict)

array([[0, 0, 0],
       [3, 0, 0],
       [0, 2, 1]])

In [ ]:
from sklearn.metrics import classification_report
print(classification_report(y_test, y_predict))

                           precision    recall  f1-score   support

        TomatoMosaicVirus       0.00      0.00      0.00         0
   TomatoSeptoriaLeafSpot       0.00      0.00      0.00         3
TomatoYellowLeafCurlVirus       1.00      0.33      0.50         3

                 accuracy                           0.17         6
                macro avg       0.33      0.11      0.17         6
             weighted avg       0.50      0.17      0.25         6



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# Model Comparison (Algorithm Exploration)

For 30 rows data

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30baris.csv")

x = df.drop(['class'], axis=1)
y = df['class']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)


from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn import model_selection
from sklearn.utils import class_weight
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import numpy as np
import pandas as pd

models = [ 
          ('Random Forest', RandomForestClassifier()),
          ('Ada Boost', AdaBoostClassifier()),
          ('Decision Tree', DecisionTreeClassifier()),
          ('KNN', KNeighborsClassifier()),
          ('SVM', SVC()), 
          ('Gaussian Naive Bayes', GaussianNB())
        ]

names = []
for name, model in models:
        clf = model.fit(x_train, y_train)
        y_pred = clf.predict(x_test)
        print(name)
        print(classification_report(y_test, y_pred))

Random Forest
                           precision    recall  f1-score   support

        TomatoMosaicVirus       0.00      0.00      0.00         0
   TomatoSeptoriaLeafSpot       0.00      0.00      0.00         3
TomatoYellowLeafCurlVirus       0.50      0.67      0.57         3

                 accuracy                           0.33         6
                macro avg       0.17      0.22      0.19         6
             weighted avg       0.25      0.33      0.29         6



/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Ada Boost
                           precision    recall  f1-score   support

        TomatoMosaicVirus       0.00      0.00      0.00         0
   TomatoSeptoriaLeafSpot       1.00      0.33      0.50         3
TomatoYellowLeafCurlVirus       0.50      0.67      0.57         3

                 accuracy                           0.50         6
                macro avg       0.50      0.33      0.36         6
             weighted avg       0.75      0.50      0.54         6

Decision Tree
                           precision    recall  f1-score   support

        TomatoMosaicVirus       0.00      0.00      0.00         0
   TomatoSeptoriaLeafSpot       1.00      0.33      0.50         3
TomatoYellowLeafCurlVirus       0.50      0.33      0.40         3

                 accuracy                           0.33         6
                macro avg       0.50      0.22      0.30         6
             weighted avg       0.75      0.33      0.45         6

KNN
                           p

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: Und

For 60 rows data

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)


from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn import model_selection
from sklearn.utils import class_weight
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import numpy as np
import pandas as pd

models = [ 
          ('Random Forest', RandomForestClassifier()),
          ('Ada Boost', AdaBoostClassifier()),
          ('Decision Tree', DecisionTreeClassifier()),
          ('KNN', KNeighborsClassifier()),
          ('SVM', SVC()), 
          ('Gaussian Naive Bayes', GaussianNB())
        ]

names = []
for name, model in models:
        clf = model.fit(x_train, y_train)
        y_pred = clf.predict(x_test)
        print(name)
        print(classification_report(y_test, y_pred))

Random Forest
                           precision    recall  f1-score   support

        TomatoMosaicVirus       1.00      0.75      0.86         4
   TomatoSeptoriaLeafSpot       1.00      0.67      0.80         3
TomatoYellowLeafCurlVirus       0.71      1.00      0.83         5

                 accuracy                           0.83        12
                macro avg       0.90      0.81      0.83        12
             weighted avg       0.88      0.83      0.83        12

Ada Boost
                           precision    recall  f1-score   support

        TomatoMosaicVirus       1.00      0.25      0.40         4
   TomatoSeptoriaLeafSpot       0.27      1.00      0.43         3
TomatoYellowLeafCurlVirus       0.00      0.00      0.00         5

                 accuracy                           0.33        12
                macro avg       0.42      0.42      0.28        12
             weighted avg       0.40      0.33      0.24        12

Decision Tree
                  

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


# Hyperparameter Tuning

##Gaussian Naive Bayes 30 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30baris.csv")

x = df.drop(['class'], axis=1)
y = df['class']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_NB = GaussianNB()

#tuning hyperparameter dengan gscv
param_grid = {'var_smoothing': np.logspace(0,-9, num=100)}
gscv = GridSearchCV(model_NB, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

Kombinasi hyperparameter terbaik :
{'var_smoothing': 0.0008111308307896872}
Dengan score validasi terbaik sebagai berikut : 
0.5900000000000001


## Gaussian Naive Bayes 60 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.naive_bayes import GaussianNB
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_NB = GaussianNB()

#tuning hyperparameter dengan gscv
param_grid = {'var_smoothing': np.logspace(0,-9, num=100)}
gscv = GridSearchCV(model_NB, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

Kombinasi hyperparameter terbaik :
{'var_smoothing': 0.0001519911082952933}
Dengan score validasi terbaik sebagai berikut : 
0.6644444444444444


## SVM 30 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30baris.csv")

x = df.drop(['class'], axis=1)
y = df['class']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_SVM = SVC()

#tuning hyperparameter dengan gscv
param_grid = {'C': [0.1, 1, 10, 100, 1000],  
              'gamma': [1, 0.1, 0.01, 0.001, 0.0001], 
              'kernel': ['linear']} 
gscv = GridSearchCV(model_SVM, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

Kombinasi hyperparameter terbaik :
{'C': 0.1, 'gamma': 1, 'kernel': 'linear'}
Dengan score validasi terbaik sebagai berikut : 
0.27999999999999997


## SVM 60 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_SVM = SVC()

#tuning hyperparameter dengan gscv
param_grid = {'C': [0.1, 1, 10, 100, 1000],  
              'gamma': [1, 0.1, 0.01, 0.001, 0.0001], 
              'kernel': ['linear']} 
gscv = GridSearchCV(model_SVM, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

Kombinasi hyperparameter terbaik :
{'C': 0.1, 'gamma': 1, 'kernel': 'linear'}
Dengan score validasi terbaik sebagai berikut : 
0.4533333333333333


## KNN 30 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_KNN = KNeighborsClassifier()

#tuning hyperparameter dengan gscv
param_grid = {'leaf_size' : list(range(1,50)), 'n_neighbors' : list(range(1,30)), 'p': [1,2]} 
gscv = GridSearchCV(model_KNN, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

Kombinasi hyperparameter terbaik :
{'leaf_size': 1, 'n_neighbors': 9, 'p': 1}
Dengan score validasi terbaik sebagai berikut : 
0.47777777777777775
Kombinasi hyperparameter terbaik :
{'leaf_size': 1, 'n_neighbors': 9, 'p': 1}
Dengan score validasi terbaik sebagai berikut : 
0.47777777777777775


## KKN 60 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_KNN = KNeighborsClassifier()

#tuning hyperparameter dengan gscv
param_grid = {'leaf_size' : list(range(1,50)), 'n_neighbors' : list(range(1,30)), 'p': [1,2]} 
gscv = GridSearchCV(model_KNN, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

Kombinasi hyperparameter terbaik :
{'leaf_size': 1, 'n_neighbors': 9, 'p': 1}
Dengan score validasi terbaik sebagai berikut : 
0.47777777777777775


## Decision Tree Classifier 30 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30baris.csv")

x = df.drop(['class'], axis=1)
y = df['class']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_DT = DecisionTreeClassifier()

#tuning hyperparameter dengan gscv
param_grid = {'criterion':['gini','entropy'], 'splitter': ['best', 'random'], 'max_depth': list(range(1,10)), 'min_samples_split': list(range (1,10)), 'min_samples_leaf':list(range(1,5))} 
gscv = GridSearchCV(model_DT, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integer 1

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integer 1

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integ

Kombinasi hyperparameter terbaik :
{'criterion': 'gini', 'max_depth': 5, 'min_samples_leaf': 2, 'min_samples_split': 3, 'splitter': 'random'}
Dengan score validasi terbaik sebagai berikut : 
0.58


##Decision Tree Classifier 60 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_DT = DecisionTreeClassifier()

#tuning hyperparameter dengan gscv
param_grid = {'criterion':['gini','entropy'], 'splitter': ['best', 'random'], 'max_depth': list(range(1,10)), 'min_samples_split': list(range (1,10)), 'min_samples_leaf':list(range(1,5))} 
gscv = GridSearchCV(model_DT, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integer 1

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integer 1

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integ

Kombinasi hyperparameter terbaik :
{'criterion': 'gini', 'max_depth': 7, 'min_samples_leaf': 3, 'min_samples_split': 5, 'splitter': 'random'}
Dengan score validasi terbaik sebagai berikut : 
0.6266666666666667


## Random Forest Classifier 30 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30baris.csv")

x = df.drop(['class'], axis=1)
y = df['class']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_RF = RandomForestClassifier()

#tuning hyperparameter dengan gscv
param_grid = {'criterion':['gini','entropy'],  'max_depth': list(range(1,10)), 'min_samples_split': list(range (1,10)), 'min_samples_leaf':list(range(1,5))} 
gscv = GridSearchCV(model_RF, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integer 1

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integer 1

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integ

Kombinasi hyperparameter terbaik :
{'criterion': 'gini', 'max_depth': 4, 'min_samples_leaf': 1, 'min_samples_split': 2}
Dengan score validasi terbaik sebagai berikut : 
0.74


## Random Forest Classifier 60 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_RF = RandomForestClassifier()

#tuning hyperparameter dengan gscv
param_grid = {'criterion':['gini','entropy'],  'max_depth': list(range(1,10)), 'min_samples_split': list(range (1,10)), 'min_samples_leaf':list(range(1,5))} 
gscv = GridSearchCV(model_RF, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integer 1

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integer 1

  FitFailedWarning)
/usr/local/lib/python3.6/dist-packages/sklearn/model_selection/_validation.py:536: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
ValueError: min_samples_split must be an integer greater than 1 or a float in (0.0, 1.0]; got the integ

Kombinasi hyperparameter terbaik :
{'criterion': 'entropy', 'max_depth': 3, 'min_samples_leaf': 1, 'min_samples_split': 7}
Dengan score validasi terbaik sebagai berikut : 
0.6711111111111111


## Ada Boost 30 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30baris.csv")

x = df.drop(['class'], axis=1)
y = df['class']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.ensemble import AdaBoostClassifier
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_AB = AdaBoostClassifier()

#tuning hyperparameter dengan gscv
param_grid = {'n_estimators': [100,200,300]} 
gscv = GridSearchCV(model_AB, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

Kombinasi hyperparameter terbaik :
{'n_estimators': 100}
Dengan score validasi terbaik sebagai berikut : 
0.38


## Ada Boost 60 rows

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)

# Code here
from sklearn.ensemble import AdaBoostClassifier
from sklearn.model_selection import GridSearchCV
import numpy as np

#menggunakan model DTR
model_AB = AdaBoostClassifier()

#tuning hyperparameter dengan gscv
param_grid = {'n_estimators': [100,200,300]} 
gscv = GridSearchCV(model_AB, param_grid=param_grid, scoring='accuracy', cv=5)
gscv.fit(x_train,y_train)

best_param = gscv.best_params_
print('Kombinasi hyperparameter terbaik :')
print(best_param)
best_score = gscv.best_score_
print('Dengan score validasi terbaik sebagai berikut : ')
print(best_score)

Kombinasi hyperparameter terbaik :
{'n_estimators': 200}
Dengan score validasi terbaik sebagai berikut : 
0.45777777777777773


# Model Comparison with the result of Hyperparameter Tuning

## 30 Rows Data

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA30baris.csv")

x = df.drop(['class'], axis=1)
y = df['class']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)


from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn import model_selection
from sklearn.utils import class_weight
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import numpy as np
import pandas as pd

models = [ 
          ('Random Forest', RandomForestClassifier(criterion= 'gini', max_depth= 4, min_samples_leaf= 1, min_samples_split= 2)),
          ('Ada Boost', AdaBoostClassifier(n_estimators= 100)),
          ('Decision Tree', DecisionTreeClassifier(criterion = 'gini', max_depth= 5, min_samples_leaf= 2, min_samples_split= 3, splitter= 'random')),
          ('KNN', KNeighborsClassifier(leaf_size= 1, n_neighbors= 9, p= 1)),
          ('SVM', SVC(C= 0.1, gamma= 1, kernel= 'linear')), 
          ('Gaussian Naive Bayes', GaussianNB(var_smoothing = 0.0008111308307896872))
        ]

names = []
for name, model in models:
        clf = model.fit(x_train, y_train)
        y_pred = clf.predict(x_test)
        print(name)
        print(classification_report(y_test, y_pred))

Random Forest
                           precision    recall  f1-score   support

   TomatoSeptoriaLeafSpot       1.00      0.33      0.50         3
TomatoYellowLeafCurlVirus       0.60      1.00      0.75         3

                 accuracy                           0.67         6
                macro avg       0.80      0.67      0.62         6
             weighted avg       0.80      0.67      0.62         6

Ada Boost
                           precision    recall  f1-score   support

   TomatoSeptoriaLeafSpot       1.00      0.33      0.50         3
TomatoYellowLeafCurlVirus       0.60      1.00      0.75         3

                 accuracy                           0.67         6
                macro avg       0.80      0.67      0.62         6
             weighted avg       0.80      0.67      0.62         6

Decision Tree
                           precision    recall  f1-score   support

        TomatoMosaicVirus       0.00      0.00      0.00         0
   TomatoSeptoria

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


## 60 Rows Data

In [ ]:
import pandas as pd
import numpy as np
df = pd.read_csv("PCA60baris.csv")

x = df.drop(['label'], axis=1)
y = df['label']

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=123)


from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn import model_selection
from sklearn.utils import class_weight
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
import numpy as np
import pandas as pd

models = [ 
          ('Random Forest', RandomForestClassifier(criterion= 'entropy', max_depth= 3, min_samples_leaf= 1, min_samples_split= 7)),
          ('Ada Boost', AdaBoostClassifier(n_estimators= 200)),
          ('Decision Tree', DecisionTreeClassifier(criterion= 'gini', max_depth= 7, min_samples_leaf= 3, min_samples_split= 5, splitter= 'random')),
          ('KNN', KNeighborsClassifier(leaf_size= 1, n_neighbors= 9, p= 1)),
          ('SVM', SVC(C= 0.1, gamma= 1, kernel= 'linear')), 
          ('Gaussian Naive Bayes', GaussianNB(var_smoothing = 0.0001519911082952933))
          ]

names = []
for name, model in models:
        clf = model.fit(x_train, y_train)
        y_pred = clf.predict(x_test)
        print(name)
        print(classification_report(y_test, y_pred))

Random Forest
                           precision    recall  f1-score   support

        TomatoMosaicVirus       1.00      0.50      0.67         4
   TomatoSeptoriaLeafSpot       1.00      0.67      0.80         3
TomatoYellowLeafCurlVirus       0.62      1.00      0.77         5

                 accuracy                           0.75        12
                macro avg       0.88      0.72      0.75        12
             weighted avg       0.84      0.75      0.74        12

Ada Boost
                           precision    recall  f1-score   support

        TomatoMosaicVirus       1.00      0.25      0.40         4
   TomatoSeptoriaLeafSpot       0.27      1.00      0.43         3
TomatoYellowLeafCurlVirus       0.00      0.00      0.00         5

                 accuracy                           0.33        12
                macro avg       0.42      0.42      0.28        12
             weighted avg       0.40      0.33      0.24        12

Decision Tree
                  

/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.6/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


SVM
                           precision    recall  f1-score   support

        TomatoMosaicVirus       0.60      0.75      0.67         4
   TomatoSeptoriaLeafSpot       0.00      0.00      0.00         3
TomatoYellowLeafCurlVirus       0.33      0.20      0.25         5

                 accuracy                           0.33        12
                macro avg       0.31      0.32      0.31        12
             weighted avg       0.34      0.33      0.33        12

Gaussian Naive Bayes
                           precision    recall  f1-score   support

        TomatoMosaicVirus       1.00      0.75      0.86         4
   TomatoSeptoriaLeafSpot       0.50      0.67      0.57         3
TomatoYellowLeafCurlVirus       0.80      0.80      0.80         5

                 accuracy                           0.75        12
                macro avg       0.77      0.74      0.74        12
             weighted avg       0.79      0.75      0.76        12



#Conclusion

From the experiments to improve performance on the type of disease data, it can be concluded that:
1. In accordance with the hypothesis at the beginning, in this case increasing the number of data train can be improved classifier performance, as evidenced by the initial accuracy of 33% to
83% 
2. In this case adding PC features degrades the performance of the classifier, evidenced
with an initial accuracy of 33% to 17% for data with 30 images and from 83% to 58% for the data with 60 images
3. In this case, performing feature scaling does not increase or decrease classifier performance, so it is known that the data does not have a far range and normally distributed
4. In this case, changing the model algorithm does not improve the performance classifier significantly. With 30 images of data, the Ada Boost model is better less performance than Naïve Bayes (accuracy to 50%). While the model Naïve Bayes and the Random Forest Classifier have the same accuracy with the data 30 images (33%) and 60 images (83%)
5. In this case, perform hyperparameter tuning of the data with 30 images improve classifier performance quite well, but on data with 60. Image accuracy was found to be lower than the un-tuned algorithm hyperparameter